qiskit_experiments/tomography/tomography_experiment.py

```diff
@@ -0,0 +1,307 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Quantum Tomography experiment
+"""
+
+import copy
+from typing import Union, Optional, Iterable, List, Tuple
+from itertools import product
+import numpy as np
+from qiskit.circuit import QuantumCircuit, Instruction
+from qiskit.circuit.library import Permutation
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+import qiskit.quantum_info as qi
+from qiskit_experiments.base_experiment import BaseExperiment, Options
+from .basis import BaseTomographyMeasurementBasis, BaseTomographyPreparationBasis
+from .tomography_analysis import TomographyAnalysis
+
+
+class TomographyExperiment(BaseExperiment):
+    """Base experiment for quantum state and process tomography"""
+
+    __analysis_class__ = TomographyAnalysis
+
+    @classmethod
+    def _default_experiment_options(cls):
+        return Options(basis_indices=None)
+
+    def __init__(
+        self,
+        circuit: Union[QuantumCircuit, Instruction, BaseOperator],
+        measurement_basis: Optional[BaseTomographyMeasurementBasis] = None,
+        measurement_qubits: Optional[Iterable[int]] = None,
+        preparation_basis: Optional[BaseTomographyPreparationBasis] = None,
+        preparation_qubits: Optional[Iterable[int]] = None,
+        basis_indices: Optional[Iterable[Tuple[List[int], List[int]]]] = None,
+        qubits: Optional[Iterable[int]] = None,
+    ):
+        """Initialize a state tomography experiment.
+
+        Args:
+            circuit: the quantum process circuit. If not a quantum circuit
+                it must be a class that can be appended to a quantum circuit.
+            measurement_basis: Tomography basis for measurements.
+            measurement_qubits: Optional, the qubits to be measured. These should refer
+                to the logical qubits in the state circuit.
+            preparation_basis: Tomography basis for measurements.
+            preparation_qubits: Optional, the qubits to be prepared. These should refer
+                to the logical qubits in the process circuit.
+            basis_indices: Optional, the basis elements to be measured. If None
+                All basis elements will be measured.
+            qubits: Optional, the physical qubits for the initial state circuit.
```

Comment:
What happens if number of `measurement` or `preparation_qubits` is grater than `qubits`? This should fail but is this captured properly? I think qubit index validation function is necessary.